### DISCUSSION TITLE
MicroPython external C modules - C++ and threading

### DISCUSSION THREAD

**Discussion by AutoPilotjn** [micropython/micropython#10493] [STM32 / Pyboard]
Hi,
I need the threading mechanism, therefore CFLAGS_EXTRA="-DMICROPY_PY_THREAD=1".
The C example works fine, the add_ints() can be imported and called with success from python. 
However, while compiling the cpp example, I get:
```CC /Users/src/cmodules/cppexample/examplemodule.c
CXX /Users/src/cmodules/cppexample/example.cpp
CC ../../py/mpstate.c
CC ../../py/nlr.c
In file included from ../../py/mpthread.h:38,
                 from ../../py/mpstate.h:32,
                 from ../../py/runtime.h:29,
                 from /Users/src/cmodules/cppexample/examplemodule.h:2,
                 from /Users/src/cmodules/cppexample/example.cpp:2:
./mpthreadport.h: In function '_mp_state_thread_t* mp_thread_get_state()':
./mpthreadport.h:40:32: error: invalid conversion from 'void*' to '_mp_state_thread_t*' [-fpermissive]
   40 |     return pyb_thread_get_local();
      |            ~~~~~~~~~~~~~~~~~~~~^~
      |                                |
      |                                void*
CC ../../py/nlrx86.c
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [build-NUCLEO_F767ZI/cppexample/example.o] Error 1
make: *** Waiting for unfinished jobs....
```
The make command:

> make -j4 BOARD=NUCLEO_F767ZI CFLAGS_EXTRA=-DMICROPY_PY_THREAD=1 FROZEN_MANIFEST=/Users/upython-1.19.1/ports/stm32/boards/manifest_NUCLEO_F767ZI_Auto_Pilot_motor_target_custom.py USER_C_MODULES=/Users/src/cmodules

I have seen this answer: https://github.com/orgs/micropython/discussions/10023#discussioncomment-4234225, but I need the threading module.
Thanks in advance for help.

**Comment by stinos** [comment]
In C one doesn't need explicit casts, but C++ does, hence the error. In other words the MicroPython code here is not C++-compatible. Can you change the code in the mentioned file into
```
static inline struct _mp_state_thread_t *mp_thread_get_state(void) {
    return (mp_state_thread_t *) pyb_thread_get_local();
}
```
and try again?

**Reply by AutoPilotjn** [reply]
With:
```
static inline struct _mp_state_thread_t *mp_thread_get_state(void) {
    return (mp_state_thread_t *) pyb_thread_get_local();
}
```
throws:
```
CC /Users/src/cmodules/cppexample/examplemodule.c
CXX /Users/src/cmodules/cppexample/example.cpp
CC ../../py/mpstate.c
In file included from ../../py/mpthread.h:38,
                 from ../../py/mpstate.h:32,
                 from ../../py/runtime.h:29,
                 from /Users/src/cmodules/cppexample/examplemodule.h:2,
                 from /Users/src/cmodules/cppexample/examplemodule.c:1:
./mpthreadport.h: In function 'mp_thread_get_state':
./mpthreadport.h:45:13: error: 'mp_state_thread_t' undeclared (first use in this function); did you mean '_mp_state_thread_t'?
   45 |     return (mp_state_thread_t *) pyb_thread_get_local();
      |             ^~~~~~~~~~~~~~~~~
      |             _mp_state_thread_t
./mpthreadport.h:45:13: note: each undeclared identifier is reported only once for each function it appears in
./mpthreadport.h:45:32: error: expected expression before ')' token
   45 |     return (mp_state_thread_t *) pyb_thread_get_local();
      |                                ^
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
```
I tried with \_mp_state_thread_t (with leading '\_'):
```
static inline struct _mp_state_thread_t *mp_thread_get_state(void) {
    return (_mp_state_thread_t *) pyb_thread_get_local();
}
```
throws:
```
./mpthreadport.h: In function 'mp_thread_get_state':
./mpthreadport.h:45:13: error: '_mp_state_thread_t' undeclared (first use in this function)
   45 |     return (_mp_state_thread_t *) pyb_thread_get_local();
      |             ^~~~~~~~~~~~~~~~~~
./mpthreadport.h:45:13: note: each undeclared identifier is reported only once for each function it appears in
./mpthreadport.h:45:33: error: expected expression before ')' token
   45 |     return (_mp_state_thread_t *) pyb_thread_get_local();
      |                                 ^
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
```
I also added an include of mpstate.h in mpthreadport.h, where mp_state_thread_t is defined, no chance: same errors.
```
// file mpthreadport.h
#include "py/mpstate.h"
```

**Comment by stinos** [comment]
> I tried with _mp_state_thread_t (with leading '_'):

Yes sorry my mistake, typo, that should be indeed `_mp_state_thread_t` with an underscore i.e. it should match the return value. That being said it should probably match completely, so 

```
static inline struct _mp_state_thread_t *mp_thread_get_state(void) {
    return (struct _mp_state_thread_t *) pyb_thread_get_local();
}
```

**Reply by AutoPilotjn** [reply]
Thanks, I confirm that:
```
static inline struct _mp_state_thread_t *mp_thread_get_state(void) {
    return (struct _mp_state_thread_t *) pyb_thread_get_local();
}
```
gets compiled successfully and the C++ test function:
```
import cppexample
cppexample.cppfunc(1, 3)
```
is imported and called with success from python side.

**Reply by stinos** [reply]
Ok great! Can you confirm this still works for the C module as well (it probably does)? If you want create a pull request so this can go in the codebase otherwise let me know then I can do it.

**Comment by AutoPilotjn** [comment]
> Can you confirm this still works for the C module

Yes, it works with both C and C++

> If you want create a pull request so this can go in the codebase otherwise let me know then I can do it

I would like to do it but I am not familiar with the PR process, however I would be glad to be guided so that I can learn how to do it !

**Comment by stinos** [comment]
It's a bit hard to find a good and complete explanation apparently, but this should get you started: glance over https://github.com/micropython/micropython/blob/master/CONTRIBUTING.md (not all super relevant for a one-line change, but still worth reading) then check the instructions here for example: https://gist.github.com/Chaser324/ce0505fbed06b947d962 (until 'Submitting', rest is not relevant).

**Reply by AutoPilotjn** [reply]
Thanks, will do. Added to my todo list ....
